sdks/java/core/src/main/java/org/apache/beam/sdk/io/UnboundedSource.java

```diff
@@ -110,8 +110,22 @@ public boolean requiresDeduping() {
      * <p>For example, this could be sending acknowledgement requests to an external
      * data source such as Pub/Sub.
      *
-     * <p>This may be called from any thread, potentially at the same time as calls to the
-     * {@code UnboundedReader} that created it.
+     * <p>Note that:
+     * <ul>
+     * <li>This finalize method may be called from any thread, potentially at the same time as
+     * calls to the {@link UnboundedReader} it was created from.
+     * <li>Checkpoints will be finalized in the same order they are created from the
+     * {@link UnboundedReader}.
+     * <li>It is possible for multiple checkpoints created from the same {@link UnboundedReader}
+     * to be 'in-flight' (ie not yet finalized) simultaneously.
+     * <li>It is possible for a checkpoint to never be finalized.
+     * <li>If this call throws an exception then the entire checkpoint will be abandoned and the
+     * reader restarted from an earlier, successfully-finalized checkpoint.
+     * <li>The system ensures that if a checkpoint fails for any reason then no later checkpoint
+     * is allowed to be finalized without the reader first being restarted.
+     * <li>It is not safe to assume the {@link UnboundedReader} from which this checkpoint was
+     * created still exists at the time this method is called.
```

Comment:
Add that in addition to the final comment, which is actually saying that the reader may have been swapped in between the checkpoint being taken and the finalize getting called even in normal operation.